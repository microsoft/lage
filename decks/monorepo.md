---
marp: true
---

## JS Monorepos: a Study of Engineering Systems

kchau@microsoft.com

---

- A practical introduction of JS Monorepos
- What problems exists for engineering systems
- Best solutions we have discovered or built

---

## Tools

Almost all of our tools are written in node.js

- frontend developers can participate in maintaining these tools
- very low friction in publishing and consuming public libraries
- enormous and open ecosystem of re-usable packages

---

# Tech stack of a Node.js tool chain

---

## V8

- C++, parses & runs JS
- Garbage collector
- Bindings for JS to access C++ libraries

---

## Node.js

- modularity with require() and module.exports
- event loop that allows single-threaded JS to have async capability
- a standard library (e.g. for I/O)

---

## Package manager (npm, yarn)

- a package is a folder with a `package.json` file
- registry as a glorified CDN
- authed API for publishing packages
- CLI tool to download package and their dependencies
- CLI to run npm "lifecycle" scripts (e.g.: build, test, bundle, lint)

---

## Node resolution algorithm

```js
// located in /project/path/main.js
require("mymodule");
```

Search paths:

- /project/path/**node_modules**/mymodule/**index.js**
- /project/**node_modules**/mymodule/**index.js**
- /**node_modules**/mymodule/**index.js**
- /**node_modules**/mymodule/**index.js**
- \$HOME/.node_modules/mymodule/**index.js**
- \$HOME/.node_libraries/mymodule/**index.js**
- \$PREFIX/lib/node/mymodule/**index.js**

---

## Monorepos

A single git repo that contains many related packages:

```
root:
  package.json
  packages/
    app/
      package.json
    businesslogic/
      package.json
    ui-components/
      package.json
    utils/
      package.json
```

---

## Importing Code

Importing internal and external code are consistent:

```js
import React from "react";
import { Button } from "ui-components";
```

---

# Monorepo Management Stack

---

## 1. Workspace-enabled package manager

- **Installs externals dependencies** for the repository
- **Links internal packages** to satisfy the node resolution algorithm
- Optional: **hoisting**, **strictness** enforcement (phantom deps)
- On the market: `yarn`, `pnpm`, `rush`, `lerna + npm`

---

## Problem 1.1: Doppelgangers

Dealing with transitive deps of different versions

```
a/
  node_modules/
    b/
      node_modules/
        react/
          package.json  <-- react@2.0.0
    c/
      node_modules/
        react/
          package.json  <-- react@2.0.0
    react/
      package.json      <-- react@1.0.0
```

> more info: https://rushjs.io/pages/advanced/npm_doppelgangers/

---

## Problem 1.2: Phantom Dependencies

`yarn` v1 will "hoist" all the deps up to the top level

```
node_modules/
  react/
    package.json        <-- react@1.0.0
b/
  package.json          <-- does not depend on react
a/
  package.json          <-- depend react
```

Now modules inside "b" can **accidentally** be allowed to depend on react

> more info: https://rushjs.io/pages/advanced/phantom_deps/

---

## Solutions

- `yarn` v2,
- `midgard-yarn` (for perf, v1 behavior)
- `pnpm` workspace
- `npm` v7+
- No silver bullet, sorry!

---

## 2. Task scheduler & runner

- **Runs npm scripts** for all packages
- **Optimize** task run speeds at the dev machine and CI
- Optionally in **topological** order or in **parallel**
- On the market: `lerna`, `wsrun`, `rush`, `pnpm recursive`, `lage`

---

## Problem Statement

Optimizes package tasks in a monorepo for a single machine

---

## Solution

A task runner that is:

- **Open sourced**
  - easily shared, public development demands **polish**
  - **easily contributed** to by many groups
- Works with all workspace implementations
- Easy setup
- Minimize idle CPU cores
- **Sublinear increase** in build time per package

---

_... doesn't exist out there, so we built one..._

# Lage

> _v. to make (Norwegian); pr. LAH-geh_

- Open sourced: https://github.com/microsoft/lage
- Easy to integrate with existing codebase
- Scales up with pipelining
- Scales out with caching and scoping
- Contributions from: OXO, FAST, ODSP, Bohemia (Fluid Exp)

---

## How does it work?

https://microsoft.github.io/lage/guide/levels.html

---

## How to try it at home?

https://microsoft.github.io/lage/guide/getting-started.html

1. npm scripts (build, test, lint) are at package level
2. `npx lage init`

   - creates a `lage.config.js` - configure it
   - adds `lage` as a dep

3. `yarn lage build` or `npm run lage build`

---

## 3. Package publish tool

- Automated management of **semver** based on a change description
- **Validation** of description of changes
- **Synchronize versions** between npm registry and git repository
- Automated management of **changelog**
- On the market: `rush`, `lerna`, `semantic-release`, `beachball`

---

## Solution

Use an automated package versioning manager

---

1. Make a new branch, edit code
2. Run a CLI tool to create a "change file"

```
> beachball change


Please describe the changes for: beachball
? Change type Â» - Use arrow-keys. Return to submit.
>   Patch      - bug fixes; no backwards incompatible changes.
    Minor      - small feature; backwards compatible changes.
    None       - this change does not affect the published package in any way.
```

3. Push branch & create a PR with complete with change file

---

## Change Files

> [An example PR](https://github.com/microsoft/beachball/pull/406/files#diff-cfc65aa360b77c3648a75d921c3876e1)

```json
{
  "type": "minor",
  "comment": "Adds the ability to create and publish canary packages",
  "packageName": "beachball",
  "email": "kchau@microsoft.com",
  "dependentChangeType": "patch",
  "date": "2020-09-11T23:37:59.115Z"
}
```

---

## Publishing to npm registry and push to git origin

> [An example publish](https://github.com/microsoft/beachball/runs/1104357745?check_suite_focus=true)

```
> beachball publish

...
Publishing - beachball@1.36.0
publish command: publish --registry https://registry.npmjs.org/ --tag latest --loglevel warn --//registry.npmjs.org/:_authToken=***
Published!
...
Pushing to https://github.com/microsoft/beachball
POST git-receive-pack (1116 bytes)
remote:
remote: GitHub found 6 vulnerabilities on microsoft/beachball's default branch (2 high, 2 moderate, 2 low). To find out more, visit:
remote:      https://github.com/microsoft/beachball/network/alerts
remote:
To https://github.com/microsoft/beachball
   72fce75..f9fa782  HEAD -> master
 * [new tag]         beachball_v1.36.0 -> beachball_v1.36.0
updating local tracking ref 'refs/remotes/origin/master'
```

---

Where to find everything we mentioned here:

- Flywheel team (OXO web engineering): <flywheel-team@microsoft.com>
- [midgard-yarn](https://www.npmjs.com/package/midgard-yarn) - yarn, but faster
- lage [repo](https://github.com/microsoft/lage), [docs](https://microsoft.github.io/lage)
- beachball [repo](https://github.com/microsoft/beachball), [docs](https://microsoft.github.io/beachball)
- One JavaScript wiki: https://aka.ms/1js

---

# Appendix

---

## Node.js modularity

The current standard is called "CommonJS"

- 1 JS file is a "module"
- modules can import and export code

---

## Imports and Exports

```js
// speech.js
module.exports = {
    sayHello: (name) => console.log(`hello ${name}`);
}

// main.js
const { sayHello } = require('./speech');
sayHello("humans!")
```

---

## Module state

require() caches the module exports as well as the state

```js
// count.js - Keeping count
let count = 1;
module.exports.inc = () => count++;

// main.js
const { inc } = require("./count");
console.log(inc()); // displays: 1
console.log(inc()); // displays: 2
```

> advanced mode: you can clear this cache to reset state & "hot reload" the module
