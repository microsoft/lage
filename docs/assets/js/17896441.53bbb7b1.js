"use strict";(self.webpackChunk_lage_run_docs=self.webpackChunk_lage_run_docs||[]).push([[401],{1828:(e,a,t)=>{t.d(a,{A:()=>n});var s=t(6540),c=t(4300);c.o.initialize({startOnLoad:!0});const n=e=>{let{chart:a}=e;return(0,s.useEffect)((()=>{c.o.contentLoaded()}),[]),s.createElement("div",{className:"mermaid"},a)}}}]);