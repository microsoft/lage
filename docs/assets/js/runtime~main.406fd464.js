(()=>{"use strict";var e,t,r,a,c,o={},d={};function f(e){var t=d[e];if(void 0!==t)return t.exports;var r=d[e]={id:e,loaded:!1,exports:{}};return o[e].call(r.exports,r,r.exports,f),r.loaded=!0,r.exports}f.m=o,f.c=d,e=[],f.O=(t,r,a,c)=>{if(!r){var o=1/0;for(b=0;b<e.length;b++){r=e[b][0],a=e[b][1],c=e[b][2];for(var d=!0,n=0;n<r.length;n++)(!1&c||o>=c)&&Object.keys(f.O).every((e=>f.O[e](r[n])))?r.splice(n--,1):(d=!1,c<o&&(o=c));if(d){e.splice(b--,1);var i=a();void 0!==i&&(t=i)}}return t}c=c||0;for(var b=e.length;b>0&&e[b-1][2]>c;b--)e[b]=e[b-1];e[b]=[r,a,c]},f.n=e=>{var t=e&&e.__esModule?()=>e.default:()=>e;return f.d(t,{a:t}),t},r=Object.getPrototypeOf?e=>Object.getPrototypeOf(e):e=>e.__proto__,f.t=function(e,a){if(1&a&&(e=this(e)),8&a)return e;if("object"==typeof e&&e){if(4&a&&e.__esModule)return e;if(16&a&&"function"==typeof e.then)return e}var c=Object.create(null);f.r(c);var o={};t=t||[null,r({}),r([]),r(r)];for(var d=2&a&&e;"object"==typeof d&&!~t.indexOf(d);d=r(d))Object.getOwnPropertyNames(d).forEach((t=>o[t]=()=>e[t]));return o.default=()=>e,f.d(c,o),c},f.d=(e,t)=>{for(var r in t)f.o(t,r)&&!f.o(e,r)&&Object.defineProperty(e,r,{enumerable:!0,get:t[r]})},f.f={},f.e=e=>Promise.all(Object.keys(f.f).reduce(((t,r)=>(f.f[r](e,t),t)),[])),f.u=e=>"assets/js/"+({11:"444aa3ef",36:"07693f84",151:"55960ee5",170:"24b81a0c",229:"27bceccc",257:"d30d811c",279:"df203c0f",287:"562a88f0",342:"f49ec8ac",401:"17896441",435:"edc893e9",447:"d3232d1d",471:"fef45042",476:"8b80ea90",571:"1356c51b",581:"935f2afb",583:"1df93b7f",674:"86f2ceba",714:"1be78505",754:"75a6a595",787:"3720c009",801:"62ec45dc",811:"a67dcc77",814:"72e14192",828:"91c16ba2",841:"6c867a5f",843:"2fb5d090",910:"31bbbc31",921:"85438e7d"}[e]||e)+"."+{11:"4f67a185",36:"7c4febb0",55:"5d40b1f1",113:"2c6e5ad6",151:"bf9717c2",170:"89457de4",229:"8b7ec60b",257:"a16a6de8",279:"f9af434a",287:"fb11f366",342:"016c036c",343:"5c8ef3ab",401:"53bbb7b1",435:"224070c2",447:"49c8d035",471:"46d759ef",476:"af3bf588",477:"07aafa51",571:"e705b1ed",581:"b26026d4",583:"964e5ce2",674:"7c8f5244",714:"7a8b8af3",754:"eb7c84ee",774:"d6207c2e",787:"90193422",801:"1a8bce4e",811:"b9919066",814:"07528205",828:"fb6197bd",841:"786a1162",843:"f97a7e78",910:"db9870be",921:"f905f69e",996:"4d3097d1"}[e]+".js",f.miniCssF=e=>{},f.g=function(){if("object"==typeof globalThis)return globalThis;try{return this||new Function("return this")()}catch(e){if("object"==typeof window)return window}}(),f.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),a={},c="@lage-run/docs:",f.l=(e,t,r,o)=>{if(a[e])a[e].push(t);else{var d,n;if(void 0!==r)for(var i=document.getElementsByTagName("script"),b=0;b<i.length;b++){var u=i[b];if(u.getAttribute("src")==e||u.getAttribute("data-webpack")==c+r){d=u;break}}d||(n=!0,(d=document.createElement("script")).charset="utf-8",d.timeout=120,f.nc&&d.setAttribute("nonce",f.nc),d.setAttribute("data-webpack",c+r),d.src=e),a[e]=[t];var l=(t,r)=>{d.onerror=d.onload=null,clearTimeout(s);var c=a[e];if(delete a[e],d.parentNode&&d.parentNode.removeChild(d),c&&c.forEach((e=>e(r))),t)return t(r)},s=setTimeout(l.bind(null,void 0,{type:"timeout",target:d}),12e4);d.onerror=l.bind(null,d.onerror),d.onload=l.bind(null,d.onload),n&&document.head.appendChild(d)}},f.r=e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},f.p="/lage/",f.gca=function(e){return e={17896441:"401","444aa3ef":"11","07693f84":"36","55960ee5":"151","24b81a0c":"170","27bceccc":"229",d30d811c:"257",df203c0f:"279","562a88f0":"287",f49ec8ac:"342",edc893e9:"435",d3232d1d:"447",fef45042:"471","8b80ea90":"476","1356c51b":"571","935f2afb":"581","1df93b7f":"583","86f2ceba":"674","1be78505":"714","75a6a595":"754","3720c009":"787","62ec45dc":"801",a67dcc77:"811","72e14192":"814","91c16ba2":"828","6c867a5f":"841","2fb5d090":"843","31bbbc31":"910","85438e7d":"921"}[e]||e,f.p+f.u(e)},(()=>{var e={354:0,869:0};f.f.j=(t,r)=>{var a=f.o(e,t)?e[t]:void 0;if(0!==a)if(a)r.push(a[2]);else if(/^(354|869)$/.test(t))e[t]=0;else{var c=new Promise(((r,c)=>a=e[t]=[r,c]));r.push(a[2]=c);var o=f.p+f.u(t),d=new Error;f.l(o,(r=>{if(f.o(e,t)&&(0!==(a=e[t])&&(e[t]=void 0),a)){var c=r&&("load"===r.type?"missing":r.type),o=r&&r.target&&r.target.src;d.message="Loading chunk "+t+" failed.\n("+c+": "+o+")",d.name="ChunkLoadError",d.type=c,d.request=o,a[1](d)}}),"chunk-"+t,t)}},f.O.j=t=>0===e[t];var t=(t,r)=>{var a,c,o=r[0],d=r[1],n=r[2],i=0;if(o.some((t=>0!==e[t]))){for(a in d)f.o(d,a)&&(f.m[a]=d[a]);if(n)var b=n(f)}for(t&&t(r);i<o.length;i++)c=o[i],f.o(e,c)&&e[c]&&e[c][0](),e[c]=0;return f.O(b)},r=self.webpackChunk_lage_run_docs=self.webpackChunk_lage_run_docs||[];r.forEach(t.bind(null,0)),r.push=t.bind(null,r.push.bind(r))})()})();