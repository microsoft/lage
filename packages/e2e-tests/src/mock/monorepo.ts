import os from "os";
import fs from "fs";
import path from "path";
import execa from "execa";

export class Monorepo {
  static tmpdir = os.tmpdir();

  root: string;
  nodeModulesPath: string;
  lagePath: string;

  constructor(private name: string) {
    this.root = fs.mkdtempSync(path.join(Monorepo.tmpdir, `lage-monorepo-${name}-`));
    this.nodeModulesPath = path.join(this.root, "node_modules");
    this.lagePath = path.join(this.nodeModulesPath, "@lage-run", "cli");
  }

  init() {
    const options = { cwd: this.root };
    execa.sync("git", ["init"], options);
    execa.sync("git", ["config", "user.email", "you@example.com"], options);
    execa.sync("git", ["config", "user.name", "test user"], options);
    execa.sync("git", ["config", "commit.gpgsign", "false"], options);
    this.generateRepoFiles();
  }

  install() {
    if (!fs.existsSync(this.nodeModulesPath)) {
      fs.mkdirSync(this.nodeModulesPath, { recursive: true });
    }

    // pretends to perform a npm install of lage
    if (!fs.existsSync(this.lagePath)) {
      fs.mkdirSync(path.dirname(this.lagePath), { recursive: true });
      fs.symlinkSync(path.join(__dirname, "..", "..", "..", "cli"), this.lagePath, "junction");
    }
  }

  /**
   * Simulates a "yarn" call by linking internal packages and generates a yarn.lock file
   */
  linkPackages() {
    const pkgs = fs.readdirSync(path.join(this.root, "packages"));

    if (!fs.existsSync(this.nodeModulesPath)) {
      fs.mkdirSync(this.nodeModulesPath, { recursive: true });
    }

    let yarnYaml = `# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n# yarn lockfile v1\n`;

    for (const pkg of pkgs) {
      fs.symlinkSync(path.join(this.root, "packages", pkg), path.join(this.nodeModulesPath, pkg), "junction");

      const pkgJson = JSON.parse(fs.readFileSync(path.join(this.root, "packages", pkg, "package.json"), "utf-8"));
      const deps = pkgJson.dependencies;

      yarnYaml += `"${pkg}@^${pkgJson.version}":\n  version "${pkgJson.version}"\n`;

      if (deps) {
        yarnYaml += `  dependencies:`;
        for (const dep of Object.keys(deps)) {
          yarnYaml += `    "${dep}" "0.1.0"`;
        }
      }
    }

    this.commitFiles({ "yarn.lock": yarnYaml });
  }

  generateRepoFiles() {
    const lagePathBin = `${this.lagePath}/lib/cli.js`;

    this.commitFiles({
      "package.json": {
        name: this.name,
        version: "0.1.0",
        private: true,
        workspaces: ["packages/*"],
        scripts: {
          bundle: `node "${lagePathBin}" bundle --reporter json --log-level silly`,
          transpile: `node "${lagePathBin}" transpile --reporter json --log-level silly`,
          build: `node "${lagePathBin}" build --reporter json --log-level silly`,
          writeInfo: `node "${lagePathBin}" info`,
          test: `node "${lagePathBin}" test --reporter json --log-level silly`,
          lint: `node "${lagePathBin}" lint --reporter json --log-level silly`,
          clear: `node "${lagePathBin}" cache --clear --reporter json --log-level silly`,
        },
        devDependencies: {
          lage: path.resolve(__dirname, "..", ".."),
        },
      },
      "lage.config.js": `module.exports = {
        pipeline: {
          build: ['^build'],
          test: ['build'],
          lint: []
        }
      };`,
      ".gitignore": "node_modules",
    });
  }

  setLageConfig(contents: string) {
    this.commitFiles({
      "lage.config.js": contents,
    });
  }

  addPackage(name: string, internalDeps: string[] = [], scripts?: { [script: string]: string }) {
    return this.commitFiles({
      [`packages/${name}/build.js`]: `console.log('building ${name}');`,
      [`packages/${name}/test.js`]: `console.log('building ${name}');`,
      [`packages/${name}/lint.js`]: `console.log('linting ${name}');`,
      [`packages/${name}/package.json`]: {
        name,
        version: "0.1.0",
        scripts: scripts || {
          build: "node ./build.js",
          test: "node ./test.js",
          lint: "node ./lint.js",
        },
        dependencies: {
          ...(internalDeps &&
            internalDeps.reduce((deps, dep) => {
              return { ...deps, [dep]: "*" };
            }, {})),
        },
      },
    });
  }

  clone(origin: string) {
    return execa.sync("git", ["clone", origin], { cwd: this.root });
  }

  push(origin: string, branch: string) {
    return execa.sync("git", ["push", origin, branch], { cwd: this.root });
  }

  commitFiles(files: { [name: string]: string | Record<string, unknown> }, options: { executable?: boolean } = {}) {
    for (const [file, contents] of Object.entries(files)) {
      let out = "";
      if (typeof contents !== "string") {
        out = JSON.stringify(contents, null, 2);
      } else {
        out = contents;
      }

      const fullPath = path.join(this.root, file);

      if (!fs.existsSync(path.dirname(fullPath))) {
        fs.mkdirSync(path.dirname(fullPath), { recursive: true });
      }

      fs.writeFileSync(fullPath, out);

      if (options.executable) {
        fs.chmodSync(path.join(this.root, file), fs.constants.S_IXUSR | fs.constants.S_IRUSR | fs.constants.S_IROTH);
      }
    }

    execa.sync("git", ["add", ...Object.keys(files)], {
      cwd: this.root,
    });

    execa.sync("git", ["commit", "-m", "commit files"], { cwd: this.root });
  }

  run(command: string, args?: string[], silent?: boolean) {
    return execa.sync("yarn", [...(silent === true ? ["--silent"] : []), command, ...(args || [])], {
      cwd: this.root,
    });
  }

  cleanup() {
    fs.rmSync(this.root, { recursive: true });
  }
}
