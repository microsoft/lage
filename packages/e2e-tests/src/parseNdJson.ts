import type { TargetStatus } from "@lage-run/scheduler-types";

export function parseNdJson(ndjson: string) {
  const entries = ndjson.substr(ndjson.indexOf("{"));

  return entries.split("\n").map((line) => {
    try {
      return JSON.parse(line);
    } catch (e) {
      return {};
    }
  });
}

export function filterEntry(data: any, pkg: string, task: string, status: TargetStatus) {
  return data?.target?.packageName === pkg && data?.target?.task === task && data.status === status;
}
