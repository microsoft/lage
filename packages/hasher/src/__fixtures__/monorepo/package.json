{
  "name": "monorepo",
  "license": "MIT",
  "version": "0.1.0",
  "workspaces": {
    "packages": [
      "packages/*"
    ]
  }
}
