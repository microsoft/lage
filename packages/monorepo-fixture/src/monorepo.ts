import * as os from "os";
import * as fs from "fs/promises";
import { existsSync, mkdtempSync } from "fs";
import * as path from "path";
import execa from "execa";

export class Monorepo {
  static tmpdir = os.tmpdir();

  root: string;

  get nodeModulesPath() {
    return path.join(this.root, "node_modules");
  }

  constructor(private name: string) {
    this.root = mkdtempSync(path.join(Monorepo.tmpdir, `monorepo-fixture-${name}-`));
  }

  async init(fixturePath?: string) {
    const options = { cwd: this.root };
    const cwd = this.root;
    await execa("git", ["init"], options);
    await execa("git", ["config", "user.email", "you@example.com"], options);
    await execa("git", ["config", "user.name", "test user"], options);
    await execa("git", ["config", "commit.gpgsign", "false"], options);

    if (fixturePath) {
      await fs.mkdir(this.root, { recursive: true });
      await fs.cp(fixturePath, this.root, { recursive: true });

      execa.sync("git", ["add", "."], { cwd });
      execa.sync("git", ["commit", "-m", "test"], { cwd });
    } else {
      await this.generateRepoFiles();
    }
  }

  async install() {
    if (!existsSync(this.nodeModulesPath)) {
      await fs.mkdir(this.nodeModulesPath, { recursive: true });
    }

    // pretends to perform a npm install of lage
    const lagePath = path.join(this.nodeModulesPath, "lage");

    if (!existsSync(lagePath)) {
      await fs.symlink(path.join(__dirname, "..", ".."), lagePath, "junction");
    }
  }

  /**
   * Simulates a "yarn" call by linking internal packages and generates a yarn.lock file
   */
  async linkPackages() {
    const pkgs = await fs.readdir(path.join(this.root, "packages"));

    if (!existsSync(this.nodeModulesPath)) {
      await fs.mkdir(this.nodeModulesPath, { recursive: true });
    }

    let yarnYaml = `# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n# yarn lockfile v1\n`;

    for (const pkg of pkgs) {
      await fs.symlink(path.join(this.root, "packages", pkg), path.join(this.nodeModulesPath, pkg), "junction");

      const pkgJson = JSON.parse(await fs.readFile(path.join(this.root, "packages", pkg, "package.json"), "utf-8"));
      const deps = pkgJson.dependencies;

      yarnYaml += `"${pkg}@^${pkgJson.version}":\n  version "${pkgJson.version}"\n`;

      if (deps) {
        yarnYaml += `  dependencies:`;
        for (const dep of Object.keys(deps)) {
          yarnYaml += `    "${dep}" "0.1.0"`;
        }
      }
    }

    await this.commitFiles({ "yarn.lock": yarnYaml });
  }

  async generateRepoFiles() {
    const lagePath = path.join(this.nodeModulesPath, "lage/lib/cli");

    await this.commitFiles({
      "package.json": {
        name: this.name,
        version: "0.1.0",
        private: true,
        workspaces: ["packages/*"],
        scripts: {
          bundle: `node "${lagePath}" bundle --reporter json --log-level silly`,
          transpile: `node "${lagePath}" transpile --reporter json --log-level silly`,
          build: `node "${lagePath}" build --reporter json --log-level silly`,
          writeInfo: `node "${lagePath}" info`,
          test: `node "${lagePath}" test --reporter json --log-level silly`,
          lint: `node "${lagePath}" lint --reporter json --log-level silly`,
          clear: `node "${lagePath}" cache --clear --reporter json --log-level silly`,
        },
        devDependencies: {
          lage: path.resolve(__dirname, "..", ".."),
        },
      },
      "lage.config.js": `module.exports = {
        pipeline: {
          build: ['^build'],
          test: ['build'],
          lint: []
        }
      };`,
      ".gitignore": "node_modules",
    });
  }

  async setLageConfig(contents: string) {
    await this.commitFiles({
      "lage.config.js": contents,
    });
  }

  async addPackage(name: string, internalDeps: string[] = [], scripts?: { [script: string]: string }) {
    return await this.commitFiles({
      [`packages/${name}/build.js`]: `console.log('building ${name}');`,
      [`packages/${name}/test.js`]: `console.log('building ${name}');`,
      [`packages/${name}/lint.js`]: `console.log('linting ${name}');`,
      [`packages/${name}/package.json`]: {
        name,
        version: "0.1.0",
        scripts: scripts || {
          build: "node ./build.js",
          test: "node ./test.js",
          lint: "node ./lint.js",
        },
        dependencies: {
          ...(internalDeps &&
            internalDeps.reduce((deps, dep) => {
              return { ...deps, [dep]: "*" };
            }, {})),
        },
      },
    });
  }

  clone(origin: string) {
    return execa("git", ["clone", origin], { cwd: this.root });
  }

  push(origin: string, branch: string) {
    return execa("git", ["push", origin, branch], { cwd: this.root });
  }

  async writeFiles(files: { [file: string]: string | object }, options: { executable?: boolean } = {}) {
    for (const [file, contents] of Object.entries(files)) {
      let out = "";
      if (typeof contents !== "string") {
        out = JSON.stringify(contents, null, 2);
      } else {
        out = contents;
      }

      const fullPath = path.join(this.root, file);

      if (!existsSync(path.dirname(fullPath))) {
        await fs.mkdir(path.dirname(fullPath), { recursive: true });
      }

      await fs.writeFile(fullPath, out);

      if (options.executable) {
        await fs.chmod(path.join(this.root, file), 0o755);
      }
    }
  }

  async readFiles(files: string[]) {
    const contents = {};
    for (const file of files) {
      const fullPath = path.join(this.root, file);
      if (!existsSync(fullPath)) {
        throw new Error(`File ${fullPath} does not exist`);
      }
      contents[file] = await fs.readFile(fullPath, "utf-8");
    }

    return contents;
  }

  async commitFiles(files: { [name: string]: string | object }, options: { executable?: boolean } = {}) {
    await this.writeFiles(files, options);
    await execa("git", ["add", "--", ...Object.keys(files)], {
      cwd: this.root,
    });
    await execa("git", ["commit", "-m", "commit files"], { cwd: this.root });
  }

  run(command: string, args?: string[], silent?: boolean) {
    return execa("yarn", [...(silent === true ? ["--silent"] : []), command, ...(args || [])], {
      cwd: this.root,
    });
  }

  cleanup() {
    return fs.rm(this.root, { recursive: true });
  }
}
