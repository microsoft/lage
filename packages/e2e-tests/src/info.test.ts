import { Monorepo } from "./mock/monorepo.js";
import { filterEntry, parseNdJson } from "./parseNdJson.js";

describe("info command", () => {
  it("basic info test case", () => {
    const repo = new Monorepo("basics-info");

    repo.init();
    repo.addPackage("a", ["b"]);
    repo.addPackage("b");

    repo.install();

    const results = repo.run("test", ["--info"]);
    const output = results.stdout + results.stderr;
    const jsonOutput = parseNdJson(output);

    expect(jsonOutput.map((entry) => ({ ...entry, cwd: `/some/path/to/${entry.packageName}` }))).toMatchSnapshot();

    repo.cleanup();
  });

  it("scoped info test case", () => {
    const repo = new Monorepo("scoped-info");

    repo.init();
    repo.addPackage("a", ["b"]);
    repo.addPackage("b");

    repo.install();

    const results = repo.run("test", ["--info", "--to", "b"]);
    const output = results.stdout + results.stderr;
    const jsonOutput = parseNdJson(output);

    expect(jsonOutput.map((entry) => ({ ...entry, cwd: `/some/path/to/${entry.packageName}` }))).toMatchSnapshot();

    repo.cleanup();
  });
});
