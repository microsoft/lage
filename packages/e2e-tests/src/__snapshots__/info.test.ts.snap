// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`info command basic info test case 1`] = `
[
  {
    "cache": true,
    "cwd": "/some/path/to/a",
    "depSpecs": [
      "build",
    ],
    "dependencies": [
      "a#build",
    ],
    "dependents": [],
    "hash": "47c840184f599036508d2fe6970e53e5241ac73f",
    "id": "a#test",
    "label": "a - test",
    "options": {},
    "packageName": "a",
    "priority": 0,
    "task": "test",
    "weight": 1,
  },
  {
    "cache": true,
    "cwd": "/some/path/to/b",
    "depSpecs": [
      "build",
    ],
    "dependencies": [
      "b#build",
    ],
    "dependents": [],
    "hash": "5a9536691284014e65dd9fffd8a60552c0a5d43a",
    "id": "b#test",
    "label": "b - test",
    "options": {},
    "packageName": "b",
    "priority": 0,
    "task": "test",
    "weight": 1,
  },
  {
    "cache": true,
    "cwd": "/some/path/to/a",
    "depSpecs": [
      "^build",
    ],
    "dependencies": [
      "b#build",
    ],
    "dependents": [
      "a#test",
    ],
    "hash": "ed53172d402c3e6a65320e54a4e376bab9caa4e4",
    "id": "a#build",
    "label": "a - build",
    "options": {},
    "packageName": "a",
    "priority": 0,
    "task": "build",
    "weight": 1,
  },
  {
    "cache": true,
    "cwd": "/some/path/to/b",
    "depSpecs": [
      "^build",
    ],
    "dependencies": [],
    "dependents": [
      "b#test",
      "a#build",
    ],
    "hash": "95ce59ba59bc59830855743a779d2cb50d069c33",
    "id": "b#build",
    "label": "b - build",
    "options": {},
    "packageName": "b",
    "priority": 0,
    "task": "build",
    "weight": 1,
  },
]
`;

exports[`info command scoped info test case 1`] = `
[
  {
    "cache": true,
    "cwd": "/some/path/to/b",
    "depSpecs": [
      "build",
    ],
    "dependencies": [
      "b#build",
    ],
    "dependents": [],
    "hash": "5a9536691284014e65dd9fffd8a60552c0a5d43a",
    "id": "b#test",
    "label": "b - test",
    "options": {},
    "packageName": "b",
    "priority": 0,
    "task": "test",
    "weight": 1,
  },
  {
    "cache": true,
    "cwd": "/some/path/to/b",
    "depSpecs": [
      "^build",
    ],
    "dependencies": [],
    "dependents": [
      "b#test",
    ],
    "hash": "95ce59ba59bc59830855743a779d2cb50d069c33",
    "id": "b#build",
    "label": "b - build",
    "options": {},
    "packageName": "b",
    "priority": 0,
    "task": "build",
    "weight": 1,
  },
]
`;
